### PR TITLE
Push down some expressions to the storage layer

### PR DESCRIPTION
**Pull request by jievince**
Associated with these two PR:
https://github.com/vesoft-inc/nebula-storage/pull/386
https://github.com/vesoft-inc/nebula-common/pull/448

### CODE REVIEW

**Comment by bright-starry-sky** [comment]
Please add test cases for lookup and match.

File: src/visitor/ExtractFilterExprVisitor.h [hunk 1]
```diff
@@ -50,6 +50,7 @@ class ExtractFilterExprVisitor final : public ExprVisitorImpl {
 
     bool canBePushed_{true};
     std::unique_ptr<Expression> remainedExpr_;
+    std::string *innerVar_;  // used by predicate, reduce and list comprehension
```
Comment (yixinglu) [inline src/visitor/ExtractFilterExprVisitor.h:53]: Where to use this variable?

**Review by yixinglu — COMMENTED** [review]
Add some tests which could not be pushed down storage

File: src/visitor/ExtractFilterExprVisitor.h (hunk 1 above)
Comment (jievince) [reply src/visitor/ExtractFilterExprVisitor.h:53]: sorry, it's unused and should be deleted

**Comment by jievince** [comment]
> Add some tests which could not be pushed down storage

I think if the filter can't be pushed down, the `filter` node and `GN` node  will not be next to each other in the generated execution plan

**Comment by jievince** [comment]
> Please add test cases for lookup and match.

But `IndexScan` node doesn't accept a filter which has a list comprehension expression
```
(root@nebula) [nba]> lookup on player where  player.age in [v in [5,6,7,8] | v + 90]
[ERROR (-12)]: SemanticError: Expression (player.age IN [v IN [5,6,7,8] | (v+90)]) not supported yet
```

File: tests/tck/features/optimizer/PushFilterDownGetNbrsRule.feature [hunk 2]
```diff
@@ -62,6 +62,35 @@ Feature: Push Filter down GetNeighbors rule
       | Project      | 1            |                                 |
       | GetNeighbors | 2            | filter: (serve.start_year>2005) |
       | Start        |              |                                 |
+    When profiling query:
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushFilterDownGetNbrsRule.feature:65]: Could you add some test cases like this:
```
$var=go from "Tim Duncan" over like yield like._dst as dst,$$.player.age as age;
go from $var.dst over like where like.likeness IN [v IN [1,3] WHERE v+$var.age > 0
```

File: src/visitor/ExtractFilterExprVisitor.cpp [hunk 3]
```diff
@@ -21,8 +21,12 @@ void ExtractFilterExprVisitor::visit(UUIDExpression *) {
...
 }
 
-void ExtractFilterExprVisitor::visit(VariableExpression *) {
-    canBePushed_ = false;
+void ExtractFilterExprVisitor::visit(VariableExpression *expr) {
+    if (static_cast<VariableExpression *>(expr)->isInner()) {
+        canBePushed_ = true;
+    } else {
+        canBePushed_ = false;
+    }
```
Comment (yixinglu) [inline src/visitor/ExtractFilterExprVisitor.cpp:29]:
```cpp
canBePushed_ = expr->isInner();
```

**Review by yixinglu — COMMENTED** [review]
Is it better to name the variable as `internal`, `local` or `temporary` than `inner`?

File: tests/tck/features/optimizer/PushFilterDownGetNbrsRule.feature (hunk 2 above)
Comment (jievince) [reply tests/tck/features/optimizer/PushFilterDownGetNbrsRule.feature:65]: But how to profile a query with a `$var`?

File: src/visitor/ExtractFilterExprVisitor.cpp (hunk 3 above)
Comment (jievince) [reply src/visitor/ExtractFilterExprVisitor.cpp:29]: ok

**Review by czpmango — APPROVED** [review]

**Pull request merged**